src/functional/ffront/foast_passes/closure_var_folding.py

```diff
@@ -0,0 +1,79 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2022, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either vervisit_Constantsion 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from typing import Any
+
+import functional.ffront.field_operator_ast as foast
+from eve import NodeTranslator, traits
+from eve.utils import FrozenNamespace
+
+
+class ClosureVarFolding(NodeTranslator, traits.VisitorWithSymbolTableTrait):
+    """
+    Replace `Name` nodes that refer to closure variable by `Constant` nodes.
+
+    `Attribute` nodes are also replaced by a `Constant` nodes with the value of the
+     closure variable's corresponding attribute. This is executed recursively for
+     attributes of attributes.
+    """
+
+    closure_vars: dict[str, Any]
+    _current_closure_vars: list[foast.Symbol]
+
+    def __init__(self, closure_vars: dict[str, Any]):
+        self.closure_vars = closure_vars
+
+    @classmethod
+    def apply(cls, node: foast.FieldOperator, closure_vars: dict[str, Any]):
```

Comment:
```suggestion
    def apply(cls, node: T, closure_vars: dict[str, Any]) -> T:
```
where `T` is a TypeVar with constraint `foast.FieldOperator | foast.ScanOperator | foast.FunctionDefinition` or simply
```suggestion
    def apply(cls, node: foast.LocatedNode, closure_vars: dict[str, Any]) -> foast.LocatedNode:
```

Reply:
I used `LocatedNode` for simplicity.